### PR TITLE
Assign Collections Randomly :game_die:

### LINKED ISSUE #135: Assign Collections Randomly
**Is your feature request related to a problem? Please describe.**
We have to optimise merkle proof computation.

**Describe the solution you'd like**
- Assign Jobs to stakers randomly solves this problem.
- Number of jobs assigned to each staker needs to be decided by governance parameter

### LINKED ISSUE #572: Resolve attack possibility of dos on dispute
dispute is dep on giveSorted, which is dep no of stakers
(refer discord for ref)

### LINKED ISSUE #643: Change test filename from assetManager.js to collectionManager.js

### LINKED ISSUE #644: Inconsistencey among equivalent data types, asset values (48,64) and medians (32)
This will allow reveal for uint48 values, but wont allow anyone to propose

### LINKED ISSUE #657: Testing Assign Collections/Assets Randomly
To @GauravJain9 
Please Update followoing as per AAR all in this #649 PR
Fork my branch and make PR there

- [ ] BlockManager
- [ ] StakeManager
- [ ] VoteManager

To @abhishekvispute : Positive cases are already tested and another positive flows would be done by above

Following are neg cases we have to do

- [ ] Proposer dont reveal non assigned assets, 
- [ ] Proposer dont reveal assigned assets
- [ ] Staker uses gigantic merkle
- [ ] Dispute happens on non assigned asset of this epoch
- [ ] Delegator is able to fetch non assigned asset

### LINKED ISSUE #684: Epic: Assign Collections Randomly
Creating Epic to combine all related issues 

- [x] #135 
- [x] #657 
- [x] #644
- [x] #643
- [x] #572

### LINKED ISSUE #699: Scenarios failing for assign collections randomly
#649

### PR DESCRIPTION
**Pull request by abhishekvispute**
Epic 
Fixes #684

Child Issues 
Fixes #135 Fixes #657 Fixes #644 Fixes #643 Fixes #572 Fixes #699

## Commit
````
Commit is changed, we are going back to merkle root being committed, and are introducing random assignment
In reveal node can only reveal collections assigned to them.
Assignment is dependent on salt and secret, salt is stored in confirm block of last epoch, and can be calculated manually.
So you are right in thinking that in commit only, stakers would know their assignment
This cant be exploited as you cant know other peoples assignment given secret is "secret" :)
So cant find if you are the only one revealing some collection
Now the tree that node has to built, has to contain all the numActiveCollections (PosAwareMerkleTree Req)
But the values for the non assigned collection doesnt matter
So we suggest to node to consider votes for non-assgined to be 0 only, so they dont have to do keccakhash redundantly
Please note that this tree is PosAware, so you cant just post gigantic tree root, and pick values as per convenience in reveal

@dev Randao Removed : So there is no penalty as randao penalty.
But Random Number Service is still there, it uses salt to get random no now
Staker is only penalised for the collections they have revealed don't match with medians.
This is decided on basis of non-zero vote value.

/// Steps
/// Fetch NumActiveCollections and no of collections to assign (toAssign)
/// Calculate Salt = keccak256(abi.encodePacked(epoch, blocks[epoch].medians))
/// Find Assignments
/// Construct Tree with values for assigned and 0 for non-assigned
/// Commitment is as per
/// commitment = keccak256(abi.encodePacked(tree.root, seed))
/// where seed = keccak256(abi.encodePacked(salt, secret));

/// @dev
/// There are two cryptographic parts
/// isAssetAlloted
/// Merkle
/// Please note that if assets alloted are 2,1,1
/// a tree is constructed for 0,1,2,3,4
/// in a tree there is no repetition
/// but in a isCollectionAlloted, input array
/// you have to pass [2,1,1], here seq and repetation should be maintained

````

## Reveal 
```
In reveal, staker has to pass secret, root of tree, revealed values and seq of allocated collections
Note that for any of assigned revealed collection, the vote value cant be zero, as if would have allowed, a clever staker could have escaped penalties.
hence 
Pls note, With this PR, result value of collections, **cant be 0**.

// Input Params
// epoch, treeRevealData, secret
// treeRevealData (follwoing struct)
// struct MerkleTree {
//     Structs.AssignedAsset [] values;
//     bytes32[][] proofs;
//     bytes32 root;
// }
```

## Propose 
```
As a proposer 
Earlier your job was easy, everyone was revealing everything
Now with randomness
Its not given that every collection is going to be revealed
isElectedProposer would use salt as seed now, not randao

// Steps
// Index reveal events of stakers
// Find medain on basis of revealed value and influence
// For non revealed active collection of this epoch, use previous epoch vote value.
// Find iteration using salt as seed
```

## Dispute 
```
GiveSorted : Will happen on sorted values now, rather than sorted stakers 
Two new fresh mint disputes :)
- disputeForProposedCollectionIds : If ids passed in propose are correct
- disputeForNonAssignedCollection : If value passed for non revealed active collection in propose is correct
More details on function definition
```

## Confirm
```
Nothing changed on user front
In contracts, salt will be stored this moment in voteStorage
```

### CODE REVIEW

File: contracts/Core/storage/BlockStorage.sol [hunk 1]
```diff
@@ -18,4 +18,6 @@ contract BlockStorage {
     int8 public blockIndexToBeConfirmed; // Index in sortedProposedBlockIds
     // epoch -> blocks
     mapping(uint32 => Structs.Block) public blocks;
+
+    bytes32 public salt;
```
Comment (GauravJain9) [inline contracts/Core/storage/BlockStorage.sol:22]: This is already present in voteManager so don't need it here.

Comment (abhishekvispute) [reply contracts/Core/storage/BlockStorage.sol:22]: yep, good catch 
lets remove it

Comment (abhishekvispute) [reply contracts/Core/storage/BlockStorage.sol]: resolved

**Comment by abhishekvispute** [comment]
Features : 
- Done: Remove salt from block storage
- Done: Salt not dependant on ep-2 at all 
- Done: Store depth, que over opt
- Remove epoch from mappings, Its a question ? Save SSTORE or LOOP=>delete that mapping and create new 
- Done: Staker have to propose asset with prev value if not revealed this epoch

Testing AAR:
Done All 
Disputes Cases 
 Proposer dont reveal non assigned assets, nonAssignedDispute
 Proposer reveals non-assgined assets, dispute should not happenfrom give sorted way and from nonAssignedDispute
 Proposer dont reveal assigned assets, giveSortedDispute
Staker shouldnt be able to use gigantic merkle
Delegator is able to fetch non assigned asset

**Comment by abhishekvispute** [comment]
````
Epoch X
0, 1,2,3
1,2,3,4
100,200,300,400


Epoch X+1
0,1,2
1,2,4

No one picks four 

100, 200, 400
```

**Comment by abhishekvispute** [comment]
Old Reveal
  218210  ·     379962  ·        306540 
  
  New Reveal
  133778  ·     244778  ·        176919

**Comment by abhishekvispute** [comment]
** Proposed Solutions**

1. Confirm : O(n) always,  we loop and update the values 
2. Keep registry for ep-2 as well, O(N), will only happen when you deactivate 
3. We discard old values and reset once we deactivate the collections, O(1)

**Comment by abhishekvispute** [comment]
Use old registry

**Comment by abhishekvispute** [comment]
Abhi
Propose : Pass Ids in propose, so we have the ref, remove index to id reg, same array can be used 
Slither 
Conflicts


Gaurav :
One scenario failing

File: test/VoteManager.js [hunk 2]
```diff
@@ -802,118 +779,118 @@ describe('VoteManager', function () {
         await assertRevert(tx2, 'Incorrect Caller');
       });
 
-      it('if the revealed value is zero, staker should be able to reveal', async function () {
```
Comment (abhishekvispute) [inline test/VoteManager.js:805]: add not negative reveal case, specified above here

**Comment by abhishekvispute** [comment]
Fixed #699
Fixes #699  (if git doesn't understand past tense to pick it up :) )

File: test/VoteManager.js (hunk 2 above)
Comment (abhishekvispute) [reply test/VoteManager.js:804]: @GauravJain9

File: test/VoteManager.js [hunk 3]
```diff
@@ -981,122 +722,104 @@ describe('VoteManager', function () {
...
-            toAdd = expectedAgeAfter2.sub(prod.div(minVoteTolerance));
-            penalty = penalty.add(toAdd);
-          }
-        }
-        expectedAgeAfter2 = toBigNumber(expectedAgeAfter2).sub(penalty);
-
-        const ageAfter2 = await stakeManager.getAge(stakerIdAcc15);
-        assertBNEqual(toBigNumber(ageAfter2), (toBigNumber(ageBefore2).add(toBigNumber('10000')).sub(penalty)), 'Incorrect Penalties given');
-      });
+      // it('Correct penalties need to be given even after an asset has been deactivated', async function () {
```
Comment (abhishekvispute) [inline test/VoteManager.js:751]: Hey @GauravJain9 
Please fix this commented testcase
Following testcase can be used as reference to fix this 
``account 4 should be penalised for incorrect voting in the previous epoch but not account 2 due to asset tolerance``

File: test/VoteManager.js [hunk 4]
```diff
@@ -565,116 +477,86 @@ describe('VoteManager', function () {
...
-      it('Should not be able to reveal with incorrect secret', async function () {
-        const epoch = await getEpoch();
-        const votes = [100, 200, 300, 400, 500, 600, 700, 800, 900];
-
-        const tx = voteManager.connect(signers[7]).reveal(epoch, votes,
-          '0x727d5c9e6d18ed15ce7ac8d3cce6ec8a0e9c02481415c0823ea49d847ccb9ddc');// last digit 'd' changed to 'c' for having incorrect secret
-
-        await assertRevert(tx, 'incorrect secret/value');
-      });
+      // it('Should not be able to reveal with incorrect secret', async function () {
```
Comment (abhishekvispute) [inline test/VoteManager.js:535]: Please fix this testcase @GauravJain9

File: test/VoteManager.js [hunk 5]
```diff
@@ -208,115 +178,78 @@ describe('VoteManager', function () {
...
-        );
-
-        const tx = voteManager.connect(signers[7]).commit(epoch, commitment1);
+        // const votes = [100, 200, 300, 400, 500, 600, 700, 800, 900];
+        const secret = '0x727d5c9e6d18ed15ce7ac8d3cce6ec8a0e9c02481415c0823ea49d847ccb8ddd';
+        const tx = commit(signers[7], 0, voteManager, collectionManager, secret);
         await assertRevert(tx, 'Staker does not exist');
       });
 
       it('should not be able to reveal if length of the votes value is not same as number of active collections', async function () {
```
Comment (abhishekvispute) [inline test/VoteManager.js:187]: Please fix this testcase as well @GauravJain9

File: test/VoteManager.js (hunk 4 above)
Comment (abhishekvispute) [reply test/VoteManager.js:535]: can be done using randomReveal ? 
(have renamed that to adhocReveal)

File: contracts/Core/BlockManager.sol [hunk 6]
```diff
@@ -48,8 +48,14 @@ contract BlockManager is Initializable, BlockStorage, StateManager, BlockManager
...
     // stakers elected in higher iterations can also propose hoping that
     // stakers with lower iteration do not propose for some reason
+
+    /// @dev The IDs being passed here, are only used for disputeForNonAssignedCollection
+    /// for delegator, we have seprate registry
+    /// If user passes invalid ids, disputeForProposedCollectionIds can happen
+
     function propose(
         uint32 epoch,
+        uint16[] memory ids,
```
Comment (abhishekvispute) [inline contracts/Core/BlockManager.sol:58]: calldata

**Comment by abhishekvispute** [comment]
So if the asset was not allotted to anyone in reveal
Proposer should not propose old value of last epoch
Proposer should propose only the assigned values

**Comment by abhishekvispute** [comment]
Build is failing because of coverwalls

File: contracts/Core/BlockManager.sol (hunk 6 above)
Comment (abhishekvispute) [reply contracts/Core/BlockManager.sol:58]: memory to calldata

**Comment by abhishekvispute** [comment]
- [ ] Calldata 
- [ ] Build Optimiser runs : Global := 2000 everywhere

### New Issues
- [ ] Proposer shouldnt pass values for non assigned assets in this epoch 
- [ ] Opt: Hash collection - getActiveCollections inside create

**Comment by SkandaBhat** [comment]
Coverage has decreased. We will be tacking this on a separate PR.

**Review by hrishikeshio — APPROVED** [review]

**Review by SamAg19 — APPROVED** [review]

**Pull request merged**